src/libraries/Microsoft.Extensions.Configuration.Xml/src/XmlStreamConfigurationProvider.cs

```diff
@@ -40,61 +39,93 @@ public class XmlStreamConfigurationProvider : StreamConfigurationProvider
                 IgnoreWhitespace = true
             };
 
+            XmlConfigurationElement? root = null;
+
             using (XmlReader reader = decryptor.CreateDecryptingXmlReader(stream, readerSettings))
             {
-                var prefixStack = new Stack<string>();
-
-                SkipUntilRootElement(reader);
-
-                // We process the root element individually since it doesn't contribute to prefix
-                ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                ProcessAttributes(reader, prefixStack, data, AddAttributePair);
+                // keep track of the tree we followed to get where we are (breadcrumb style)
+                var currentPath = new Stack<XmlConfigurationElement>();
 
                 XmlNodeType preNodeType = reader.NodeType;
+
                 while (reader.Read())
                 {
                     switch (reader.NodeType)
                     {
                         case XmlNodeType.Element:
-                            prefixStack.Push(reader.LocalName);
-                            ProcessAttributes(reader, prefixStack, data, AddNamePrefix);
-                            ProcessAttributes(reader, prefixStack, data, AddAttributePair);
-
-                            // If current element is self-closing
-                            if (reader.IsEmptyElement)
                             {
-                                prefixStack.Pop();
+                                var element = new XmlConfigurationElement(reader.LocalName, GetName(reader), GetLineInfo(reader));
+
+                                XmlConfigurationElement parent = currentPath.Any() ? currentPath.Peek() : null;
+
+                                if (parent == null)
```

Comment:
We could avoid this check right? We could do something like:

```cs
if (!currentPath.Any())
{
    root = element;
}
else
{
     XmlConfigurationElement parent = currentPath.Peek();
     ...
}
```

I don't believe `currentPath.Peek()`can return null right?

Reply:
Nice find, updating.